core/services/gateway/connectionmanager.go

```diff
@@ -72,40 +117,155 @@ func (m *connectionManager) DONConnectionManager(donId string) DONConnectionMana
+	signer, err := ValidateSignature(signature, authHeader[:n-network.HandshakeSignatureLen])
+	nodeAddress = "0x" + hex.EncodeToString(signer)
+	return
+}
+
+func (m *connectionManager) newAttempt(nodeSt *nodeState, nodeAddress string, timestamp uint32) (string, []byte, error) {
+	challenge := make([]byte, m.config.AuthChallengeLen)
+	_, err := rand.Read(challenge)
+	if err != nil {
+		return "", nil, err
+	}
+	m.connAttemptsMu.Lock()
+	defer m.connAttemptsMu.Unlock()
+	m.connAttemptCounter++
+	newId := fmt.Sprintf("%s_%d", nodeAddress, m.connAttemptCounter)
+	m.connAttempts[newId] = &connAttempt{nodeState: nodeSt, nodeAddress: nodeAddress, challenge: challenge, timestamp: timestamp}
+	return newId, challenge, nil
 }
 
 func (m *connectionManager) FinalizeHandshake(attemptId string, response []byte, conn *websocket.Conn) error {
-	m.lggr.Debug("FinalizeHandshake", attemptId, response)
-	// TODO (FUN-469): implement the handshake (validate signature, add a new connection)
+	m.lggr.Debug("FinalizeHandshake attempt: ", attemptId)
+	m.connAttemptsMu.Lock()
+	attempt, ok := m.connAttempts[attemptId]
+	m.connAttemptsMu.Unlock()
+	if !ok {
+		return errors.New("connection attempt not found")
+	}
+	signer, err := ValidateSignature(response, attempt.challenge)
+	if err != nil {
+		return errors.New("invalid challenge response")
+	}
+	if attempt.nodeAddress != "0x"+hex.EncodeToString(signer) {
+		return errors.New("invalid signer")
+	}
+	attempt.nodeState.mu.Lock()
+	defer attempt.nodeState.mu.Unlock()
+	if attempt.nodeState.lastAcceptedTs >= attempt.timestamp {
+		return errors.New("timestamp too low")
+	}
+	m.lggr.Infof("Node %s connected!", attempt.nodeAddress)
+	attempt.nodeState.conn.Restart(conn)
 	return nil
 }
 
 func (m *connectionManager) AbortHandshake(attemptId string) {
-	m.lggr.Debug("AbortHandshake", attemptId)
-	// TODO (FUN-469): implement the handshake (clear cached attempt)
+	m.lggr.Debug("AbortHandshake attempt:", attemptId)
+	m.connAttemptsMu.Lock()
+	defer m.connAttemptsMu.Unlock()
+	delete(m.connAttempts, attemptId)
 }
 
 func (m *donConnectionManager) SetHandler(handler Handler) {
-	m.mu.Lock()
-	defer m.mu.Unlock()
 	m.handler = handler
 }
 
 func (m *donConnectionManager) SendToNode(ctx context.Context, nodeAddress string, msg *Message) error {
-	return nil
+	data, err := m.codec.EncodeRequest(msg)
+	if err != nil {
+		return fmt.Errorf("error encoding request for node %s: %v", nodeAddress, err)
+	}
+	return m.nodes[nodeAddress].conn.Write(ctx, websocket.BinaryMessage, data)
+}
+
+func (m *donConnectionManager) readLoop(nodeAddress string, nodeState *nodeState) {
+	for {
+		select {
+		case <-m.shutdownCh:
+			m.closeWait.Done()
+			return
+		case item := <-nodeState.conn.ReadChannel():
+			msg, err := m.codec.DecodeResponse(item.Data)
+			if err != nil {
+				m.lggr.Error("parse error when reading from node ", nodeAddress, err)
+				break
+			}
+			err = m.handler.HandleNodeMessage(context.Background(), msg, nodeAddress)
```

Comment:
I missed this Background context.. you won't be able to shutdown until one handler returns. Maybe this is another candidate to create a context from chan.